Source/Diagnostics/ComputeDiagFunctors/ParticleReductionFunctor.cpp

```diff
@@ -0,0 +1,132 @@
+
+#include "ParticleReductionFunctor.H"
+
+#include "Diagnostics/ComputeDiagFunctors/ComputeDiagFunctor.H"
+#include "Particles/MultiParticleContainer.H"
+#include "Particles/WarpXParticleContainer.H"
+#include "Utils/CoarsenIO.H"
+#include "WarpX.H"
+
+#include <AMReX_Array.H>
+#include <AMReX_BLassert.H>
+#include <AMReX_IntVect.H>
+#include <AMReX_MultiFab.H>
+#include <AMReX_REAL.H>
+
+using namespace amrex::literals;
+
+ParticleReductionFunctor::ParticleReductionFunctor (const amrex::MultiFab* mf_src, const int lev,
+        amrex::IntVect crse_ratio, const std::string fn_str, const int ispec, const int ncomp)
+    : ComputeDiagFunctor(ncomp, crse_ratio), m_lev(lev), m_ispec(ispec)
+{
+    // mf_src will not be used, let's make sure it's null.
+    AMREX_ALWAYS_ASSERT(mf_src == nullptr);
+    // Write only in one output component.
+    AMREX_ALWAYS_ASSERT(ncomp == 1);
+
+    // Allocate and compile a parser based on the input string fn_str
+    m_map_fn_parser = std::make_unique<amrex::Parser>(makeParser(
+                fn_str, {"x", "y", "z", "ux", "uy", "uz"}));
+    m_map_fn = m_map_fn_parser->compile<6>();
+}
+
+void
+ParticleReductionFunctor::operator() (amrex::MultiFab& mf_dst, const int dcomp, const int /*i_buffer*/) const
+{
+    auto& warpx = WarpX::GetInstance();
+    // Guard cell is set to 1 for generality. However, for a cell-centered
+    // output Multifab, mf_dst, the guard-cell data is not needed especially considering
+    // the operations performend in the CoarsenAndInterpolate function.
+    constexpr int ng = 1;
+    // Temporary cell-centered, multi-component MultiFab for storing particles per cell.
+    amrex::MultiFab red_mf(warpx.boxArray(m_lev), warpx.DistributionMap(m_lev), 1, ng);
+    amrex::MultiFab ppc_mf(warpx.boxArray(m_lev), warpx.DistributionMap(m_lev), 1, ng);
+    // Set value to 0, and increment the value in each cell with ppc.
+    red_mf.setVal(0._rt);
+    ppc_mf.setVal(0._rt);
+    auto& pc = warpx.GetPartContainer().GetParticleContainer(m_ispec);
+    // Copy over compiled parser function so that it can be captured by value in the lambda
+    auto map_fn = m_map_fn;
+    ParticleToMesh(pc, red_mf, m_lev,
+            [=] AMREX_GPU_DEVICE (const WarpXParticleContainer::SuperParticleType& p,
+                amrex::Array4<amrex::Real> const& out_array,
+                amrex::GpuArray<amrex::Real,AMREX_SPACEDIM> const& plo,
+                amrex::GpuArray<amrex::Real,AMREX_SPACEDIM> const& dxi)
+            {
+                // Get position in WarpX convention to use in parser. Will be different from
+                // p.pos() for 1D and 2D simulations.
+                amrex::ParticleReal xw = 0, yw = 0, zw = 0;
+                get_particle_position(p, xw, yw, zw);
+
+                // Get position in AMReX convention to calculate corresponding index.
+                // Ideally this will be replaced with the AMReX NGP interpolator
+                amrex::ParticleReal x = 0, y = 0, z = 0;
```

Comment:
Same, do these maybe need the `_rt` suffix for consistency?
```suggestion
                amrex::ParticleReal x = 0._rt, y = 0._rt, z = 0._rt;
```